### ISSUE TITLE
rustdoc: `Methods from Deref<Target = T<U>>` is imprecise (contains false positives)

### ISSUE THREAD

**Issue by ghost** [rust-lang/rust#24686]
``` rust
use std::ops::Deref;

pub struct Foo<T>(T);
impl Foo<i32> {
    pub fn get_i32(&self) -> i32 { self.0 }
}
impl Foo<u32> {
    pub fn get_u32(&self) -> u32 { self.0 }
}
pub struct Bar(Foo<i32>);
impl Deref for Bar {
    type Target = Foo<i32>;
    fn deref(&self) -> &Foo<i32> {
        &self.0
    }
}
```

The `Methods from Deref` for `Bar` will contain both `get_i32` and `get_u32`, but should only contain `get_i32`.

[Image](http://i.imgur.com/bUEO5pb.png)

**Comment by tomjakubowski** [comment]
Just wondering (to help prioritize this), is there an example of this bug in the standard library docs?

**Comment by whataloadofwhat** [comment]
I don't think there's anything which matches this particular example exactly.

The closest thing that I could find was `DerefVec` which includes all methods of `Vec` and doesn't separate them into different bounds like it does for `Vec`. So someone reading the docs for `DerefVec` *may* think that you can `dedup` on any `DerefVec<T>`, when in fact you can only do it  where `T: PartialEq`.

http://doc.rust-lang.org/nightly/collections/vec/struct.DerefVec.html

**Comment by tomjakubowski** [comment]
Ah yeah. Preserving bounds in the deref is related but distinct enough it would be worth creating a new issue for it.




As an aside: when I tried to do the "methods available through Deref" feature myself, a couple months ago, I got stuck enumerating all of these miscellaneous issues and pondering pie in the sky solutions for them. I'm so happy acrichto stepped in to ship something covering the 95% use case :-)



—

**Comment by sfackler** [comment]
I hit this trying to make a unified `Ref<T>` wrapper for rust-openssl: https://github.com/sfackler/rust-openssl/blob/ab30ad0ce76ee6fd8cc14ecff2186171c7f610e3/openssl/src/types.rs

I'll probably work around it by dropping the generic `Ref` wrapper and making a separate `FooRef` type for each `Foo`, but it's a bit unfortunate.

**Comment by lolbinarycat** [comment]
Did some more looking into this, and I think the issue is here, where we are throwing away the generic arguments of the `Target` associated type:
 https://github.com/rust-lang/rust/blob/617aad8c2e8783f6df8e5d1f8bb1e4bcdc70aa7b/src/librustdoc/html/render/mod.rs#L1417

Unfortunately, I can't find the other half, the generic arguments of the impl block that the methods are from, so I can't fix this, but I'll be adding a `known-bug` test in a PR.

**Comment by lolbinarycat** [comment]
This seems to have regressed somewhat recently, as now the two impl blocks are merged into one!

**Comment by lolbinarycat** [comment]
Is "methods from deref" a field in rustdoc-json output? if it is, and it has the same bug, this likely needs to be fixed somewhere outside of `render`.

**Comment by lolbinarycat** [comment]
Figured it out, the key function is `is_doc_subtype_of`.

**Comment by lolbinarycat** [comment]
Hmm.. it seems the sidebar is generated separately...

**Comment by lolbinarycat** [comment]
it's been improved significantly, but rustdoc still doesn't have the info to resolve type aliases and compare then for equality, I believe we need the lazy_type_alias feature first.

**Comment by retrokp** [comment]
Is this issue the reason why the [experimental](https://github.com/rust-lang/rust/issues/134915) [`ByteString`](https://doc.rust-lang.org/std/bstr/struct.ByteString.html#deref-methods-%5BT%5D) shows Methods from Deref<Target = [T]>, but not methods from Deref<Target = [u8]>?

The end result is that many methods are missing from `ByteString` rustdoc, such as `as_ascii()`, `is_ascii()` and `make_ascii_lowercase()`. I've tested with a simple program that the methods are actually there for `ByteString`. They are just not in the documentation.

[`ByteStr`](https://doc.rust-lang.org/std/bstr/struct.ByteStr.html#deref-methods-%5Bu8%5D) shows the Methods from Deref<Target = [u8]> correctly, including `as_ascii()`, `is_ascii()` and `make_ascii_lowercase()`.

Checked on Rust 1.98.1.

**Comment by lolbinarycat** [comment]
@retrokp most likely, yes.
